pkg/datastore/pipedstore.go

```diff
@@ -60,12 +107,37 @@ func (p *pipedCollection) Encode(e interface{}) (map[Shard][]byte, error) {
 		return nil, fmt.Errorf(errFmt, "type not matched")
 	}
 
-	data, err := json.Marshal(me)
+	clientShardStruct := me
+	cdata, err := json.Marshal(&clientShardStruct)
 	if err != nil {
 		return nil, fmt.Errorf(errFmt, "unable to marshal entity data")
 	}
+
+	agentShardStruct := model.Piped{
+		// Fields which must exists due to the validation check on update.
+		Id:        me.Id,
+		Name:      me.Name,
+		ProjectId: me.ProjectId,
+		CreatedAt: me.CreatedAt,
+		UpdatedAt: me.UpdatedAt,
+		// Fields which value only available in AgentShard.
+		CloudProviders: me.CloudProviders,
+		Repositories:   me.Repositories,
+		StartedAt:      me.StartedAt,
+		// Fields which be committed by Piped once at it start, after that
+		// those fields value can be updated by WebCommander so we should use
+		// those values from ClientShard with a higher priority.
+		Version:          me.Version,
```

Comment:
Version too, it belongs to only Piped commander.

Reply:
WebCommander updates the `DesiredVersion` instead of this.

Reply:
Ah, nice catch. Thanks 🙌